File: frontend/src/components/chat/providers/cursor/extractors/extensions.ts
import type { ImageResultSource } from '~/lib/imageBlocks'
import type { TodoItem } from '~/models/todo'
import { CURSOR_EXTENSION_FRAME, CURSOR_METHOD, CURSOR_SUPPLEMENT } from '~/generated/contracts/cursor-protocol'
import { isObject, pickObject, pickString } from '~/lib/jsonPick'
import { rawTodosToItems } from '~/models/todo'

/**
 * One `cursor/*` extension frame, as the worker stored it on a tool row.
 *
 * Cursor sends the frame as a JSON-RPC REQUEST, one per finished call, right after
 * the update that completed the same `toolCallId`. The worker answers it and writes
 * it into that row's supplemental content, because the frame carries fields the tool
 * call itself never states. See `cursor_extensions.go`.
 */
export interface CursorExtension {
  method: string
  params: Record<string, unknown>
}

/** Read the stored frame off one row's supplemental content, or null for every other row. */
export function cursorExtension(supplemental: Record<string, unknown> | undefined): CursorExtension | null {
  const stored = pickObject(supplemental, CURSOR_SUPPLEMENT.Extension)
  const method = pickString(stored, CURSOR_EXTENSION_FRAME.Method)
  if (!method)
    return null
  return { method, params: pickObject(stored, CURSOR_EXTENSION_FRAME.Params) ?? {} }
}

/**
 * The to-do list one `updateTodos` row shows.
 *
 * Two sources, and the stored frame wins. Cursor spells a status twice: the frame
 * normalizes it (`in_progress`), and the tool call's own `rawInput.todos` repeats the
 * protobuf enum NAME (`TODO_STATUS_IN_PROGRESS`). A row written before the worker
 * stored these frames carries the second form alone, so both are read.
 *
 * The frame's `merge` flag is deliberately not drawn. The row shows what Cursor
 * reported in that call, which is what the reader saw happen; the sidebar owns the
 * whole list, and the worker applies the same flag there.
 */
export function cursorTodoItems(extension: CursorExtension | null, input: Record<string, unknown>): TodoItem[] {
  const fromFrame = extension?.method === CURSOR_METHOD.UpdateTodos ? extension.params[CURSOR_EXTENSION_FRAME.Todos] : undefined
  if (Array.isArray(fromFrame))
    return rawTodosToItems(fromFrame)
  const raw = input.todos
  if (!Array.isArray(raw))
    return []
  return rawTodosToItems(raw.map(entry => isObject(entry) ? { ...entry, status: cursorTodoStatusWord(entry.status) } : entry))
}

/**
 * Fold Cursor's protobuf enum NAME onto the word every other list uses.
 *
 * `rawInput.todos` carries `TODO_STATUS_IN_PROGRESS`, which the shared normalizer
 * reads as `pending` -- the state it claims the least about -- so every row of an
 * older transcript drew as unstarted. A value that is already the word passes
 * through unchanged.
 */
function cursorTodoStatusWord(status: unknown): unknown {
  if (typeof status !== 'string' || !status.startsWith('TODO_STATUS_'))
    return status
  return status.slice('TODO_STATUS_'.length).toLowerCase()
}

/**
 * The images one `generateImage` row shows.
 *
 * `filePath` is where the runtime WROTE the image, which the call's own
 * `rawInput.filename` only requested. The reference images are the inputs the
 * prompt pointed at, and they are shown after the result so the reader sees the
 * output first.
 *
 * Each source carries a path and no bytes: the viewer opens the file through the
 * worker, which is the same route a Claude `Read` of an image takes.
 */
export function cursorGeneratedImages(extension: CursorExtension | null): ImageResultSource[] {
  if (extension?.method !== CURSOR_METHOD.GenerateImage)
    return []
  const description = pickString(extension.params, 'description')
  const produced = pickString(extension.params, 'filePath')
  const references = Array.isArray(extension.params.referenceImagePaths) ? extension.params.referenceImagePaths : []
  const sources: ImageResultSource[] = produced ? [{ filePath: produced, ...(description ? { description } : {}) }] : []
  for (const reference of references) {
    if (typeof reference === 'string' && reference)
      sources.push({ filePath: reference, description: 'Reference image' })
  }
  return sources
}

/**
 * What one `task` row states about the subagent that ran.
 *
 * The tool call carries the prompt, the description and the REQUESTED type. The
 * frame adds the model that answered, the runtime's own agent id and the measured
 * duration, so the row reports the run rather than the request.
 */
export function cursorTaskDetails(extension: CursorExtension | null): Record<string, unknown> {
  if (extension?.method !== CURSOR_METHOD.Task)
    return {}
  const details: Record<string, unknown> = {}
  for (const key of ['subagentType', 'model', 'agentId', 'durationMs'] as const) {
    const value = extension.params[key]
    if (value !== undefined && value !== null && value !== '')
      details[key] = value
  }
  return details
}

File: backend/internal/worker/agent/providers/cursor/extensions.go
package cursor

import (
	"encoding/json"
	"fmt"
	"log/slog"

	"github.com/leapmux/leapmux/generated/contracts"
	"github.com/leapmux/leapmux/internal/worker/agent"
	"github.com/leapmux/leapmux/internal/worker/agent/providers/acp"
	"github.com/leapmux/leapmux/internal/worker/todoevents"
)

// cursorTodoItem is one row of a cursor/update_todos frame.
//
// The three fields are the SHARED to-do vocabulary rather than Cursor's own, so no
// contract holds them and contracts.CursorExtensionParams keeps `Todos` raw.
//
// Status is Cursor's own lowercase word -- `pending`, `in_progress`, `completed` or
// `cancelled` -- which is the normalized form. The tool call's rawInput.todos spells
// the same state as the protobuf enum name (`TODO_STATUS_IN_PROGRESS`), so this frame
// is also the half that needs no second parser.
type cursorTodoItem struct {
	ID      string `json:"id"`
	Content string `json:"content"`
	Status  string `json:"status"`
}

// cursorTodoFrame is the payload of cursor/update_todos.
//
// The shared half is the GENERATED contracts.CursorExtensionParams, whose json tags
// the browser reads back out of the stored frame. `Merge` stays here, because the
// worker applies the flag and the browser deliberately draws none from it, so the
// contracts rule exempts it. The embedded `Todos` stays raw, and cursorStoredTodoFrame
// decodes it into `todos` in one more step.
type cursorTodoFrame struct {
	contracts.CursorExtensionParams
	// Merge false REPLACES the list. Cursor sends it on the first update of a turn and
	// sends true afterwards, with the rows that changed. Reading a true frame as a
	// replacement would delete every row it stayed silent about.
	Merge bool `json:"merge"`
	todos []cursorTodoItem
}

// handleCursorExtension stores one `cursor/*` extension frame on the row of its tool
// call, and reports whether the frame was one this worker knows.
//
// Cursor sends one after the tool-call update that COMPLETES the same toolCallId, on
// the same stream, so the row the worker persisted for that call is the row the frame
// describes. Each frame carries fields the tool call itself does not:
//
//   - cursor/update_todos adds `merge`, which states whether the listed rows replace
//     the whole list or join it. The tool call's own rawInput.todos carries the rows
//     with no word for that, so the flag is the one thing that makes the list usable.
//   - cursor/task adds the model, the agent id and the measured duration of a subagent
//     run. rawInput carries the prompt, the description and the requested type alone.
//   - cursor/generate_image adds the path the image was WRITTEN to. rawInput carries
//     the requested filename, which the runtime does not have to honor.
//
// It NEVER refuses the frame because the write failed. Cursor keeps no state for an
// extension request and an error reply would reach a reader as a failed turn, so the
// ack is unconditional and a failed write is a log line.
func (a *Agent) handleCursorExtension(method string, params json.RawMessage) bool {
	switch method {
	case contracts.CursorMethodUpdateTodos, contracts.CursorMethodTask, contracts.CursorMethodGenerateImage:
	default:
		return false
	}
	var frame contracts.CursorExtensionParams
	if err := json.Unmarshal(params, &frame); err != nil || frame.ToolCallID == "" {
		// A frame with no tool call identifies no row. It is not an error the reader can
		// act on, so it is a trace and not a warning.
		slog.Debug("Cursor extension frame identifies no tool call", "method", method, "error", err)
		return true
	}
	if method == contracts.CursorMethodTask {
		a.noteCursorTaskExtension(frame.ToolCallID)
	}
	if a.transcript == nil {
		// Only a test builds a Cursor agent with no transcript, and it does so to
		// exercise the ack. Persisting from here would need the transcript's own
		// single-writer discipline -- see EnrichToolSpan.
		return true
	}
	build := func(original []byte) ([]byte, error) { return cursorExtensionSupplement(original, method, params) }
	written, err := a.transcript.EnrichToolSpan(frame.ToolCallID, build)
	switch {
	case err != nil:
		slog.Warn("Store the Cursor extension frame on its tool row",
			"method", method, "tool_call_id", frame.ToolCallID, "error", err)
	case !written:
		// A false with no error is a REFUSAL, not a failure: no row carries the tool
		// call, or EnrichMessage declined a write whose PreviousRevision no longer
		// matches the row. The frame is lost either way -- a `cursor/update_todos`
		// row then draws as a plain tool call with no checklist -- and this is the
		// only line that states it, because the error branch above sees nothing.
		slog.Debug("No row took the Cursor extension frame",
			"method", method, "tool_call_id", frame.ToolCallID)
	}
	return true
}

// cursorExtensionSupplement wraps one frame for the row's supplemental content.
//
// It IDENTIFIES the frame it belongs to, like every other supplement this provider family
// stores. Without the identity keys the browser's shared gate refused the envelope --
// while the worker's own resolve took it -- so a `cursor/update_todos` row drew as a
// plain tool call with no checklist, and neither side reported anything.
func cursorExtensionSupplement(original []byte, method string, params json.RawMessage) ([]byte, error) {
	var frame map[string]json.RawMessage
	if err := json.Unmarshal(original, &frame); err != nil {
		return nil, fmt.Errorf("decode the row's own frame: %w", err)
	}
	stored, err := json.Marshal(contracts.CursorStoredExtension{Method: method, Params: params})
	if err != nil {
		return nil, fmt.Errorf("encode the stored frame: %w", err)
	}
	supplement := acp.NewToolSupplement(frame)
	supplement[contracts.CursorSupplementExtension] = stored
	encoded, err := json.Marshal(supplement)
	if err != nil {
		return nil, fmt.Errorf("encode the supplement: %w", err)
	}
	return encoded, nil
}

// ExtractTodoEvent reads Cursor's to-do list off one persisted message.
//
// The list reaches the row as a SUPPLEMENT, not in the provider's own message: the
// tool call carries the rows and the cursor/update_todos frame that follows it carries
// the `merge` flag, and a list without that flag cannot be applied. ResolveProviderData
// puts the stored frame back into the content this reads.
//
// It DELEGATES to the embedded Provider for every other message. Cursor speaks ACP,
// so an ACP plan notification still reads as one.
func (p cursorProvider) ExtractTodoEvent(spanType string, content []byte, paired func() []byte) (todoevents.Event, bool) {
	frame, ok := cursorStoredTodoFrame(content)
	if !ok {
		return p.Provider.ExtractTodoEvent(spanType, content, paired)
	}
	items := make([]todoevents.Item, 0, len(frame.todos))
	for _, todo := range frame.todos {
		if todo.ID == "" {
			continue
		}
		items = append(items, todoevents.Item{
			ID:      todo.ID,
			Content: todo.Content,
			Status:  todoevents.StatusFromProviderWord(todo.Status),
		})
	}
	// A frame that listed rows and yielded none states nothing about the list: every
	// row it carried had no id, which the loop above skips. A REPLACEMENT built from
	// it deletes every row and broadcasts an empty checklist, so a single malformed
	// frame wipes the list. The OpenCode sibling guards the same case.
	if len(frame.todos) > 0 && len(items) == 0 {
		return todoevents.Event{}, false
	}
	if frame.Merge {
		// An empty merge frame states that nothing changed, so it is not an event.
		// An empty REPLACEMENT frame is the real answer "the list is now empty".
		if len(items) == 0 {
			return todoevents.Event{}, false
		}
		return todoevents.Event{Kind: todoevents.KindMerge, Items: items}, true
	}
	return todoevents.Event{Kind: todoevents.KindSnapshot, Snapshot: items}, true
}

// cursorStoredTodoFrame reads a stored cursor/update_todos frame out of one message.
func cursorStoredTodoFrame(content []byte) (cursorTodoFrame, bool) {
	// Indexed with the generated constant rather than a struct tag, which cannot
	// hold one. Every other reader and the writer take this key from the contract,
	// so a hand-written tag here would survive a rename and silently stop matching:
	// ExtractTodoEvent would fall through to the ACP base and Cursor's to-do list
	// would stop updating, with no build error and no log line.
	var message map[string]json.RawMessage
	if json.Unmarshal(content, &message) != nil {
		return cursorTodoFrame{}, false
	}
	stored, ok := message[contracts.CursorSupplementExtension]
	if !ok {
		return cursorTodoFrame{}, false
	}
	var extension contracts.CursorStoredExtension
	if json.Unmarshal(stored, &extension) != nil || extension.Method != contracts.CursorMethodUpdateTodos {
		return cursorTodoFrame{}, false
	}
	var frame cursorTodoFrame
	if json.Unmarshal(extension.Params, &frame) != nil {
		return cursorTodoFrame{}, false
	}
	// The generated struct keeps `Todos` raw, so the rows take one more step. A frame
	// whose rows do not decode is refused here, exactly as a frame whose outer shape
	// is wrong is refused above -- ExtractTodoEvent then falls through to the ACP
	// base for both, rather than reading a malformed list as an empty one.
	if len(frame.Todos) > 0 && json.Unmarshal(frame.Todos, &frame.todos) != nil {
		return cursorTodoFrame{}, false
	}
	return frame, true
}

// ResolveProviderData gives semantic extractors the stored extension frame.
//
// It DELEGATES to the embedded Provider and then adds Cursor's own key, so an ACP
// resolution rule added later reaches Cursor too. The shared resolver copies a fixed
// set of ACP field names and cannot carry a name only Cursor writes.
func (p cursorProvider) ResolveProviderData(content agent.MessageContent) []byte {
	resolved := p.Provider.ResolveProviderData(content)
	if len(content.Supplemental) == 0 {
		return resolved
	}
	var supplement acp.ToolSupplement
	if json.Unmarshal(content.Supplemental, &supplement) != nil {
		return resolved
	}
	extension, present := supplement[contracts.CursorSupplementExtension]
	if !present {
		return resolved
	}
	var fields map[string]json.RawMessage
	if json.Unmarshal(resolved, &fields) != nil || fields == nil {
		return resolved
	}
	// The SAME gate the shared ACP resolve applies, and the same gate the browser
	// plugin applies. A supplement stored beside one frame must not reach another.
	if !supplement.IdentityMatches(fields) {
		return resolved
	}
	fields[contracts.CursorSupplementExtension] = extension
	encoded, err := json.Marshal(fields)
	if err != nil {
		return resolved
	}
	return encoded
}
